File: packages/shared/python/iopsych_contracts/role_extraction.py
"""Strict role-extraction contracts and source-evidence validation."""

from typing import Annotated, Literal, Self

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, model_validator

from .constructs import CONSTRUCT_KEYS, ConfidenceLevel, ConstructKey, Rating

NonEmptyString = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]


class RoleConstructRating(BaseModel):
    """An LLM-proposed rating that remains subject to human review."""

    model_config = ConfigDict(extra="forbid")

    key: ConstructKey
    rating: Rating
    confidence: ConfidenceLevel
    rationale: NonEmptyString
    evidence: Annotated[list[NonEmptyString], Field(min_length=1)]


class RoleExtraction(BaseModel):
    """Complete draft-only extraction output for all six constructs."""

    model_config = ConfigDict(extra="forbid")

    constructs: Annotated[
        list[RoleConstructRating],
        Field(min_length=len(CONSTRUCT_KEYS), max_length=len(CONSTRUCT_KEYS)),
    ]
    assumptions: list[NonEmptyString]
    needs_human_review: Literal[True]

    @model_validator(mode="after")
    def require_all_constructs_and_review(self) -> Self:
        """Reject duplicate/missing constructs and any attempt to bypass review."""

        observed_keys = {construct.key for construct in self.constructs}
        if observed_keys != set(CONSTRUCT_KEYS):
            raise ValueError("constructs must contain each version 1 construct exactly once")
        return self

    def validate_evidence(self, job_description: str) -> Self:
        """Require every excerpt to occur verbatim in the submitted description."""

        if not job_description.strip():
            raise ValueError("a non-empty job description is required")

        for construct in self.constructs:
            for excerpt in construct.evidence:
                if excerpt not in job_description:
                    raise ValueError(
                        f"evidence for {construct.key.value} is not present in the job description"
                    )
        return self


def parse_role_extraction(payload: object, job_description: str) -> RoleExtraction:
    """Validate both the serialized structure and its trusted source context."""

    return RoleExtraction.model_validate(payload).validate_evidence(job_description)

File: packages/shared/src/role-extraction.ts
import { z } from "zod";

import {
  ConfidenceLevelSchema,
  CONSTRUCT_KEYS,
  ConstructKeySchema,
  RatingSchema,
} from "./constructs";

export const RoleConstructRatingSchema = z.strictObject({
  key: ConstructKeySchema,
  rating: RatingSchema,
  confidence: ConfidenceLevelSchema,
  rationale: z.string().trim().min(1),
  evidence: z.array(z.string().trim().min(1)).min(1),
});
export type RoleConstructRating = z.infer<typeof RoleConstructRatingSchema>;

/** Structural LLM output contract. Context-dependent evidence checks are added below. */
export const RoleExtractionSchema = z
  .strictObject({
    constructs: z
      .array(RoleConstructRatingSchema)
      .length(CONSTRUCT_KEYS.length),
    assumptions: z.array(z.string().trim().min(1)),
    needs_human_review: z.literal(true),
  })
  .superRefine((extraction, context) => {
    const observedKeys = new Set(
      extraction.constructs.map((construct) => construct.key),
    );

    for (const key of CONSTRUCT_KEYS) {
      if (!observedKeys.has(key)) {
        context.addIssue({
          code: "custom",
          message: `Missing construct: ${key}`,
          path: ["constructs"],
        });
      }
    }
  });
export type RoleExtraction = z.infer<typeof RoleExtractionSchema>;

/**
 * Bind the structural extraction contract to its submitted job description.
 * Evidence must be a verbatim, case-sensitive substring of that source text.
 */
export function roleExtractionSchemaForJobDescription(jobDescription: string) {
  return RoleExtractionSchema.superRefine((extraction, context) => {
    extraction.constructs.forEach((construct, constructIndex) => {
      construct.evidence.forEach((excerpt, evidenceIndex) => {
        if (!jobDescription.includes(excerpt)) {
          context.addIssue({
            code: "custom",
            message:
              "Evidence must appear verbatim in the submitted job description",
            path: ["constructs", constructIndex, "evidence", evidenceIndex],
          });
        }
      });
    });
  });
}

export function parseRoleExtraction(
  payload: unknown,
  jobDescription: string,
): RoleExtraction {
  if (jobDescription.trim().length === 0) {
    throw new Error(
      "A non-empty job description is required to validate extraction evidence",
    );
  }

  return roleExtractionSchemaForJobDescription(jobDescription).parse(payload);
}
